src/tactic/has_variable_names.lean

```diff
@@ -0,0 +1,165 @@
+/-
+Copyright (c) 2020 Jannis Limperg. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jannis Limperg
+-/
+
+import tactic.core
+
+/-!
+# A tactic for type-based naming of variables
+
+When we name hypotheses or variables, we often do so in a type-directed fashion:
+a hypothesis of type `ℕ` is called `n` or `m`; a hypothesis of type `list ℕ` is
+called `ns`; etc. This module provides a tactic,
+`tactic.typical_variable_names`, which looks up typical variable names for a
+given type. Typical variable names are registered by giving an instance of the
+type class `has_variable_names`.
+-/
+
+universes u v
+
+/--
+Type class for associating a type `α` with typical variable names for elements
+of `α`. See `tactic.typical_variable_names`.
+-/
+meta class has_variable_names (α : Sort u) : Type :=
+(first_name : name)
+(other_names : list name)
+
+
+namespace tactic
+
+/--
+`typical_variable_names t` obtains typical names for variables of type `t`.
+The returned list is guaranteed to be nonempty. Fails if there is no instance
+`has_typical_variable_names t`.
+
+```
+typical_variable_names `(ℕ) = [`n, `m, `o]
+```
+-/
+meta def typical_variable_names (t : expr) : tactic (list name) :=
+(do
+  first_name ← to_expr ``(has_variable_names.first_name %%t),
+  n ← eval_expr name first_name,
+  other_names ← to_expr ``(has_variable_names.other_names %%t),
+  ns ← eval_expr (list name) other_names,
+  pure $ n :: ns)
+<|> fail! "typical_variable_names: unable to get typical variable names for type {t}"
```

Comment:
How about a non-tactic function `has_variable_names.to_list : Sort u -> list name := first_name :: other_names`, so that the code in the tactic can be shortened to `(to_expr ``(has_variable_names.to_list %%t) >>= eval_expr (list name)) <|> fail`?

Reply:
This is a good idea; I'll implement it.